Add missing text issue #3539 (#3544)
OSGeo/gdal: gdal/doc/source/api/gdalogrin_python.rst

.. _gdalogrin_python:

> ================================================================================

Reviewer: You would strongly benefit from resynchronizing the outdated content of this page with swig/python/README.rst which was updated recently

Author: Fix with gdal\swig\python\README.rst

GDAL/OGR in Python
================================================================================

> This Python package and extensions are a number of tools for programming and manipulating the `GDAL <http://www.gdal.org/>`__ Geospatial Data Abstraction Library. Actually, it is two libraries -- GDAL for manipulating geospatial raster data and 

Reviewer: paragraph should be rewritten as just "This Python package and extensions are a number of tools for programming and manipulating GDAL"

Author: Fixed

OGR for manipulating geospatial vector data -- but we'll refer to the entire package as the GDAL library for the purposes of this document.

> The GDAL project (primarily Howard Butler) maintains SWIG generated Python bindings for GDAL and OGR.

Reviewer: ```suggestion
The GDAL project maintains SWIG generated Python bindings for GDAL and OGR.

```

API
---

Generally speaking the classes and methods mostly match those of the GDAL and OGR C++ classes. An automatically generated API documentation is available on `http://gdal.org/python <http://gdal.org/python>`__.
It is far from being ideal however : you will essentially get the names of available classes and functions, but few descriptions (except for OGR classes where the OGR C documentation is copied). For more details,

> look at the `GDAL API Tutorial <http://www.gdal.org/gdal_tutorial.html>`__ includes Python examples.

Reviewer: ```suggestion
look at the :ref:`tutorials <tutorials>` includes Python examples.

```

Tutorials
---------

Chris Garrard has given courses at Utah State University on "Geoprocessing with Python using Open Source GIS" (`http://www.gis.usu.edu/~chrisg/python <http://www.gis.usu.edu/~chrisg/python>`__). There are many slides, examples, test data... and homework ;-) that can
be greatly helpful for beginners with GDAL/OGR in Python.

Greg Petrie has an OGR related tutorial available at `http://cosmicproject.org/OGR <http://cosmicproject.org/OGR>`__.

A cookbook full of recipes for using the Python GDAL/OGR bindings : `http://pcjericks.github.io/py-gdalogr-cookbook/index.html <http://pcjericks.github.io/py-gdalogr-cookbook/index.html>`__

Examples
--------

> * One example of GDAL/numpy integration is found in the val_repl.py script.

Reviewer: please update the links in this paragraph to github

Author: Fixed

* An assortment of other samples are available in the Python SVN samples directory at `http://svn.osgeo.org/gdal/trunk/gdal/swig/python/samples/ <http://svn.osgeo.org/gdal/trunk/gdal/swig/python/samples/>`__ with some description in the `samples README <http://svn.osgeo.org/gdal/trunk/gdal/swig/python/samples/README>`__.
* Several `GDAL utilities <http://svn.osgeo.org/gdal/trunk/gdal/swig/python/scripts/>`__ are implemented in Python and can be useful examples.
* The majority of GDAL regression tests are written in Python. They are available at `http://svn.osgeo.org/gdal/trunk/autotest <http://svn.osgeo.org/gdal/trunk/autotest>`__

Gotchas
-------

Although GDAL's and OGR's Python bindings provide a fairly "Pythonic" wrapper around the underlying C++ code, there are several ways in which the Python bindings differ from typical Python libraries.
These differences can catch Python programmers by surprise and lead to unexpected results. These differences result from the complexity of developing a large, long-lived library while continuing to maintain

> backward compatibility. They are being addressed over time, but until they are all gone, please review this list of `Python gotchas <http://trac.osgeo.org/gdal/wiki/PythonGotchas>`__.

Reviewer: ```suggestion
backward compatibility. They are being addressed over time, but until they are all gone, please review this list of :ref:`python_gotchas`.

```

but they should be enough to get you going. Docstrings for GDAL and OSR are planned for a future release.

The History of Using GDAL/OGR in Python
---------------------------------------

Python was the first set of bindings supported by GDAL/OGR and though the bindings were generated with SWIG (1.1 series), the process was very Python specific and contained a significant
amount of hand written wrapper code. In 2005, Kevin Ruland launched an effort for a set of next generation bindings generated with SWIG (1.3 series) and supported by a variety of languages.
With GDAL 1.4.0 the various bindings became fairly mature, and for GDAL 1.5.0, the "next-generation" bindings become the default bindings. The previous, "old-generation," bindings will continue to be available
, but they will not be widely supported and no new development will be targeted at them. From the viewpoint of a user, with GDAL 1.5.0 and above, you should not have to worry very much about the distinction
between these two development efforts.

> Usage of Old-Generation Python Bindings

Reviewer: Paragraph to remove

Author: fixed

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For certain legacy applications (most notably OpenEV 1.x), it may be necessary to continue to use the old-generation Python bindings. These can be built and installed as part of a source build from ./configure:

.. code-block:: Bash

   ./configure --with-ogpython=/usr/bin/python

Numpy/Numeric
-------------

One advanced feature of the GDAL Python bindings not found in the other language bindings (C#, Perl) is integration with the Python numerical array facilities. The gdal.Dataset.ReadAsArray() method can
be used to read raster data as numerical arrays, ready to use with the Python numerical array capabilities.

These facilities have evolved somewhat over time. In the past the package was known as "Numeric" and imported using "import Numeric". A new generation is imported using "import numpy". Currently the old
generation bindings only support the older Numeric package, and the new generation bindings only support the new generation numpy package. They are mostly compatible, and by importing gdalnumeric (or osgeo.gdal_array)
you will get whichever is appropriate to the current bindings type.

Examples
~~~~~~~~

> One example of GDAL/numpy integration is found in the `val_repl.py <http://trac.osgeo.org/gdal/browser/trunk/gdal/swig/python/samples/val_repl.py>`__ script.

Reviewer: link to update to github

Author: Fixed
